From b66c0852aeb0509cff348307890ff0309d48f40f Mon Sep 17 00:00:00 2001
From: Dimitar Dimitrov <dinuxbg@gmail.com>
Date: Mon, 28 Jul 2014 23:10:08 +0300
Subject: [PATCH 2/4] HACK: Limit LBBO/SBBO bursts to 20 regs

Until gcc/genextract.c is fixed to work with gargantuan
machine-generated insn patterns, limit PRU bursts to 20 registers.

Signed-off-by: Dimitar Dimitrov <dinuxbg@gmail.com>
---
 gcc/config/pru/pru-ldst-multiple.md | 1242 -----------------------------------
 gcc/config/pru/pru-ldst-multiple.ml |    2 +-
 gcc/config/pru/pru.md               |    7 +
 3 files changed, 8 insertions(+), 1243 deletions(-)

diff --git a/gcc/config/pru/pru-ldst-multiple.md b/gcc/config/pru/pru-ldst-multiple.md
index 6d6488a..a18f091 100644
--- a/gcc/config/pru/pru-ldst-multiple.md
+++ b/gcc/config/pru/pru-ldst-multiple.md
@@ -1332,1245 +1332,3 @@
   [(set_attr "type" "ld")
    (set_attr "length" "4")])
 
-(define_insn "*lbbo_multiple_21"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 22 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 22)
-                  (const_int 80))))])]
-  "XVECLEN (operands[0], 0) == 21"
-  "lbbo\t%1, %22, 0, 84"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_21"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 22 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 22) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 21"
-  "sbbo\t%1, %22, 0, 84"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_22"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 23 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 23)
-                  (const_int 84))))])]
-  "XVECLEN (operands[0], 0) == 22"
-  "lbbo\t%1, %23, 0, 88"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_22"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 23 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 23) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 22"
-  "sbbo\t%1, %23, 0, 88"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_23"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 24 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 84))))
-     (set (match_operand:SI 23 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 24)
-                  (const_int 88))))])]
-  "XVECLEN (operands[0], 0) == 23"
-  "lbbo\t%1, %24, 0, 92"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_23"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 24 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 24) (const_int 88)))
-          (match_operand:SI 23 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 23"
-  "sbbo\t%1, %24, 0, 92"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_24"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 25 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 84))))
-     (set (match_operand:SI 23 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 88))))
-     (set (match_operand:SI 24 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 25)
-                  (const_int 92))))])]
-  "XVECLEN (operands[0], 0) == 24"
-  "lbbo\t%1, %25, 0, 96"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_24"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 25 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 88)))
-          (match_operand:SI 23 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 25) (const_int 92)))
-          (match_operand:SI 24 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 24"
-  "sbbo\t%1, %25, 0, 96"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_25"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 26 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 84))))
-     (set (match_operand:SI 23 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 88))))
-     (set (match_operand:SI 24 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 92))))
-     (set (match_operand:SI 25 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 26)
-                  (const_int 96))))])]
-  "XVECLEN (operands[0], 0) == 25"
-  "lbbo\t%1, %26, 0, 100"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_25"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 26 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 88)))
-          (match_operand:SI 23 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 92)))
-          (match_operand:SI 24 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 26) (const_int 96)))
-          (match_operand:SI 25 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 25"
-  "sbbo\t%1, %26, 0, 100"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_26"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 27 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 84))))
-     (set (match_operand:SI 23 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 88))))
-     (set (match_operand:SI 24 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 92))))
-     (set (match_operand:SI 25 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 96))))
-     (set (match_operand:SI 26 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 27)
-                  (const_int 100))))])]
-  "XVECLEN (operands[0], 0) == 26"
-  "lbbo\t%1, %27, 0, 104"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_26"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 27 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 88)))
-          (match_operand:SI 23 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 92)))
-          (match_operand:SI 24 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 96)))
-          (match_operand:SI 25 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 27) (const_int 100)))
-          (match_operand:SI 26 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 26"
-  "sbbo\t%1, %27, 0, 104"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_27"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 28 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 84))))
-     (set (match_operand:SI 23 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 88))))
-     (set (match_operand:SI 24 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 92))))
-     (set (match_operand:SI 25 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 96))))
-     (set (match_operand:SI 26 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 100))))
-     (set (match_operand:SI 27 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 28)
-                  (const_int 104))))])]
-  "XVECLEN (operands[0], 0) == 27"
-  "lbbo\t%1, %28, 0, 108"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_27"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 28 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 88)))
-          (match_operand:SI 23 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 92)))
-          (match_operand:SI 24 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 96)))
-          (match_operand:SI 25 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 100)))
-          (match_operand:SI 26 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 28) (const_int 104)))
-          (match_operand:SI 27 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 27"
-  "sbbo\t%1, %28, 0, 108"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_28"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 29 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 84))))
-     (set (match_operand:SI 23 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 88))))
-     (set (match_operand:SI 24 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 92))))
-     (set (match_operand:SI 25 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 96))))
-     (set (match_operand:SI 26 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 100))))
-     (set (match_operand:SI 27 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 104))))
-     (set (match_operand:SI 28 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 29)
-                  (const_int 108))))])]
-  "XVECLEN (operands[0], 0) == 28"
-  "lbbo\t%1, %29, 0, 112"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_28"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 29 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 88)))
-          (match_operand:SI 23 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 92)))
-          (match_operand:SI 24 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 96)))
-          (match_operand:SI 25 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 100)))
-          (match_operand:SI 26 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 104)))
-          (match_operand:SI 27 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 29) (const_int 108)))
-          (match_operand:SI 28 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 28"
-  "sbbo\t%1, %29, 0, 112"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*lbbo_multiple_29"
-  [(match_parallel 0 "pru_load_multiple_operation"
-    [(set (match_operand:SI 1 "register_operand" "")
-          (mem:SI (match_operand:SI 30 "register_operand" "r")))
-     (set (match_operand:SI 2 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 4))))
-     (set (match_operand:SI 3 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 8))))
-     (set (match_operand:SI 4 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 12))))
-     (set (match_operand:SI 5 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 16))))
-     (set (match_operand:SI 6 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 20))))
-     (set (match_operand:SI 7 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 24))))
-     (set (match_operand:SI 8 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 28))))
-     (set (match_operand:SI 9 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 32))))
-     (set (match_operand:SI 10 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 36))))
-     (set (match_operand:SI 11 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 40))))
-     (set (match_operand:SI 12 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 44))))
-     (set (match_operand:SI 13 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 48))))
-     (set (match_operand:SI 14 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 52))))
-     (set (match_operand:SI 15 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 56))))
-     (set (match_operand:SI 16 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 60))))
-     (set (match_operand:SI 17 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 64))))
-     (set (match_operand:SI 18 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 68))))
-     (set (match_operand:SI 19 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 72))))
-     (set (match_operand:SI 20 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 76))))
-     (set (match_operand:SI 21 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 80))))
-     (set (match_operand:SI 22 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 84))))
-     (set (match_operand:SI 23 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 88))))
-     (set (match_operand:SI 24 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 92))))
-     (set (match_operand:SI 25 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 96))))
-     (set (match_operand:SI 26 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 100))))
-     (set (match_operand:SI 27 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 104))))
-     (set (match_operand:SI 28 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 108))))
-     (set (match_operand:SI 29 "register_operand" "")
-          (mem:SI (plus:SI (match_dup 30)
-                  (const_int 112))))])]
-  "XVECLEN (operands[0], 0) == 29"
-  "lbbo\t%1, %30, 0, 116"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
-(define_insn "*sbbo_multiple_29"
-  [(match_parallel 0 "pru_store_multiple_operation"
-    [(set (mem:SI (match_operand:SI 30 "register_operand" "r"))
-          (match_operand:SI 1 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 4)))
-          (match_operand:SI 2 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 8)))
-          (match_operand:SI 3 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 12)))
-          (match_operand:SI 4 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 16)))
-          (match_operand:SI 5 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 20)))
-          (match_operand:SI 6 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 24)))
-          (match_operand:SI 7 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 28)))
-          (match_operand:SI 8 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 32)))
-          (match_operand:SI 9 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 36)))
-          (match_operand:SI 10 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 40)))
-          (match_operand:SI 11 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 44)))
-          (match_operand:SI 12 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 48)))
-          (match_operand:SI 13 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 52)))
-          (match_operand:SI 14 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 56)))
-          (match_operand:SI 15 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 60)))
-          (match_operand:SI 16 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 64)))
-          (match_operand:SI 17 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 68)))
-          (match_operand:SI 18 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 72)))
-          (match_operand:SI 19 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 76)))
-          (match_operand:SI 20 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 80)))
-          (match_operand:SI 21 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 84)))
-          (match_operand:SI 22 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 88)))
-          (match_operand:SI 23 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 92)))
-          (match_operand:SI 24 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 96)))
-          (match_operand:SI 25 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 100)))
-          (match_operand:SI 26 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 104)))
-          (match_operand:SI 27 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 108)))
-          (match_operand:SI 28 "register_operand" ""))
-     (set (mem:SI (plus:SI (match_dup 30) (const_int 112)))
-          (match_operand:SI 29 "register_operand" ""))])]
-  "XVECLEN (operands[0], 0) == 29"
-  "sbbo\t%1, %30, 0, 116"
-  [(set_attr "type" "ld")
-   (set_attr "length" "4")])
-
diff --git a/gcc/config/pru/pru-ldst-multiple.ml b/gcc/config/pru/pru-ldst-multiple.ml
index cee6b77..326e8fe 100644
--- a/gcc/config/pru/pru-ldst-multiple.ml
+++ b/gcc/config/pru/pru-ldst-multiple.ml
@@ -81,7 +81,7 @@ let write_pattern_1 name ls nregs write_set_fn =
 let patterns () =
   (* The rule for load-multiple of a single register greatly simplifies
      the prologue/epilogue code. *)
-  for nregs = 1 to 29; do
+  for nregs = 1 to 20; do
     write_pattern_1 "lbbo" "load" nregs write_ldm_set;
     write_pattern_1 "sbbo" "store" nregs write_stm_set;
   done
diff --git a/gcc/config/pru/pru.md b/gcc/config/pru/pru.md
index c55aa05..8c9f745 100644
--- a/gcc/config/pru/pru.md
+++ b/gcc/config/pru/pru.md
@@ -42,6 +42,11 @@
    (FRAME_POINTER_REGNUM       33)
    (ARG_POINTER_REGNUM         34)
    (FIRST_PSEUDO_REGISTER      35)
+
+   ;; Misc
+   (MAX_XBBO_BURST_LEN	    20)	; Artificially limited by GCC - see how
+				; genextract.c uses 'a'..'z' to record
+				; "path to a vector".
   ]
 )
 
@@ -103,6 +108,7 @@
      only if at least one register. */
   if (GET_CODE (operands[2]) != CONST_INT
       || INTVAL (operands[2]) < 1
+      || INTVAL (operands[2]) > MAX_XBBO_BURST_LEN
       || GET_CODE (operands[1]) != MEM
       || GET_CODE (operands[0]) != REG
       || (REGNO (operands[0]) + INTVAL (operands[2]) - 1) > LAST_NONIO_GP_REG)
@@ -139,6 +145,7 @@
      only if at least one register. */
   if (GET_CODE (operands[2]) != CONST_INT
       || INTVAL (operands[2]) < 1
+      || INTVAL (operands[2]) > MAX_XBBO_BURST_LEN
       || GET_CODE (operands[0]) != MEM
       || GET_CODE (operands[1]) != REG
       || (REGNO (operands[1]) + INTVAL (operands[2]) - 1) > LAST_NONIO_GP_REG)
-- 
2.1.0.rc1

